src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java

```diff
@@ -0,0 +1,207 @@
+import org.spongepowered.common.data.DataProcessor;
+import org.spongepowered.common.data.manipulator.immutable.entity.ImmutableSpongePassengerData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongePassengerData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class PassengerDataProcessor implements DataProcessor<PassengerData, ImmutablePassengerData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof net.minecraft.entity.Entity;
+    }
+
+    @SuppressWarnings("unused")
+    @Override
+    public Optional<PassengerData> from(DataHolder dataHolder) {
+        if (dataHolder instanceof net.minecraft.entity.Entity) {
+            final SpongePassengerData passengerData = new SpongePassengerData();
+            final Entity vehicle = (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) vehicle;
+
+            while (currentEntity != null && currentEntity.isRiding())
+            {
+                currentEntity = currentEntity.ridingEntity;
+            }
+
+            final Entity baseVehicle = (Entity) currentEntity;
+
+            return Optional.<PassengerData>of(passengerData.setVehicle(vehicle).setBaseVehicle(baseVehicle));
+        } else {
+            return Optional.absent();
+        }
+
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataHolder dataHolder, PassengerData manipulator) {
+        if (dataHolder instanceof net.minecraft.entity.Entity) {
+
+            manipulator.set(Keys.VEHICLE, (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity);
+
+            net.minecraft.entity.Entity currentEntity = ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+
+            while (currentEntity != null && currentEntity.isRiding())
+            {
+                currentEntity = currentEntity.ridingEntity;
+            }
+
+            final Entity baseVehicle = (Entity) currentEntity;
+
+            manipulator.set(Keys.BASE_VEHICLE, baseVehicle);
+
+            return Optional.of(manipulator);
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataHolder dataHolder, PassengerData manipulator, MergeFunction overlap) {
+        if (dataHolder instanceof net.minecraft.entity.Entity) {
+            final PassengerData merged = overlap.merge(checkNotNull(manipulator).copy(), from(dataHolder).get());
+            manipulator.set(Keys.BASE_VEHICLE, merged.baseVehicle().get())
+                    .set(Keys.VEHICLE, merged.vehicle().get());
+            return Optional.of(manipulator);
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataContainer container, PassengerData passengerData) {
+        passengerData.set(Keys.BASE_VEHICLE, getData(container, Keys.BASE_VEHICLE));
```

Comment:
I'm not 100% sure we want to be deserializing a full entity here. This implies that we ARE deserializing a full `Entity` object, not just finding the next appropriate entity with the valid UUID.